### PR TITLE
Document WAI concepts & common operations

### PR DESCRIPTION
**Pull request by Michael-F-Bryan**
## Description

This updates the documentation to include core WAI concepts (records, variants, etc.) as well as some how-to guides for common problems you might face (e.g. deciding whether to use a resource or a resource).

## Context

This is part of a larger documentation push I'm making to flesh out the Wasmer SDK. Hopefully, by writing these things down, it'll be easier to plan where we want to go and enunciate why it is important.

## Checklist

- Is this a user-facing change? If so, update [`CHANGELOG.md`](https://github.com/wasmerio/wasmer-pack/blob/master/CHANGELOG.md)

### CODE REVIEW

File: doc/concepts/wai/README.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+The WebAssembly Interfaces project (WAI for short) provides a polyfill for
+passing around higher-level objects. It lets developers define their imports and
+exports in a `*.wai` file, then uses `wai-bindgen` to generate glue which
+automagically passes things around within the constraints of WebAssembly.
+
+There are four main parts to WAI:
+
+- The `*.wai` file
+- the WAI Bindgen code generator
```
Comment (dynamite-bud) [inline doc/concepts/wai/README.md:29]: **T** should be capital

File: doc/concepts/wai/README.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+The WAI Bindgen code generator consumes `*.wai` files and generates glue code
+that [the host](#the-host) can use for using functionality from a WebAssembly
+module or [the guest](#the-guest) can use for implementing that functionality.
+
+There are two primary ways users will interact with WAI Bindgen, the
+`wai-bindgen` CLI, and the `wai-bindgen-*` family of crates.
+
+The `wai-bindgen` CLI provides a command-line interface to the `wai-bindgen-*`
+crates, and is often used for once-off investigation or integration with a
+non-Rust build system.
```
Comment (dynamite-bud) [inline doc/concepts/wai/README.md:66]: Does **non-Rust** have correct capitalization ?

File: doc/concepts/wai/records.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+would be referred to as a [plain old data][pod] type.
+
+Depending on the language, records may be expressed in in different ways.
+
+| Language   | Equivalent Construct     |
+| ---------- | ------------------------ |
+| Rust       | Struct                   |
+| C          | Struct                   |
+| Python     | [Data Class][dataclass]  |
+| JavaScript | [Type alias][type-alias] |
```
Comment (dynamite-bud) [inline doc/concepts/wai/records.md:14]: You wanted to refer to TypeScript maybe

File: doc/concepts/wai/README.md (hunk 2 above)
Comment (Michael-F-Bryan) [reply doc/concepts/wai/README.md:80]: Yep. Rust is a noun, so it will be capitalised even if it has a `non-` prefix.

**Pull request merged**